Limit active object step time budget

lhofhansl opened the pull request:
Split from #6667 .

Streamlines the time budget for active objects and also introduces a budget for active objects.
This prevent active objects to taking up too much time.

Uses a floating point multiplier now to simply increase the interval instead of skipping an integer number of intervals.

With heavy mobs, this prevent them from suddenly "flying off", or from using so much time that no more blocks are generated - both seen with NSSM.

If/when this is merged, I'll rebase #6667 accordingly.

paramat commented on src/serverenvironment.cpp:
```diff
@@ -279,17 +279,17 @@ void LBMManager::applyLBMs(ServerEnvironment *env, MapBlock *block, u32 stamp)
+	for (p.Y = p0.Y - r; p.Y <= p0.Y + r; p.Y++)
+	for (p.Z = p0.Z - r; p.Z <= p0.Z + r; p.Z++) {
+		// limit to a sphere
+		if (p.getDistanceFromSQ(p0) <= r2) {
```
'r2'?

lhofhansl replied:
See the constant above. Saves a sqrt per point to check.

paramat replied:
Ugh i missed seeing that, sorry.

lhofhansl commented:
I played around with this. For mob heady worlds this improves things significantly. NSSM became playable (it was close to useless before).

sofar commented:
Casual glance, but this seems simple and more than reasonable to me.

nerzhul commented on src/serverenvironment.cpp:
```diff
@@ -1306,6 +1303,20 @@ void ServerEnvironment::step(float dtime)
+		// calculate a simple moving average
+		m_avg_ao_time = m_avg_ao_time * 0.9f + timer.stop(true) * 0.1f;
+		// allow up to 10% of the budget interval
+		const u32 max_time_ms = m_cache_ao_interval * 1000 / 10;
```
better to keep float here

lhofhansl replied:
Why is that? You're worried about the resolution when m_cache_ao_interval is very small?

nerzhul replied:
no, i just want to use a float instead of u32 for the comparison with the interval

SmallJoker commented on src/serverenvironment.cpp:
```diff
@@ -1260,27 +1253,31 @@ void ServerEnvironment::step(float dtime)
 	*/
-	{
+	if (m_active_object_interval.step(dtime, m_cache_ao_interval * m_active_object_interval_overload_skip)) {
+		float dtime = m_cache_ao_interval * m_active_object_interval_overload_skip;
```
Why do you initialize `dtime` a second time here?

lhofhansl replied:
Yeah. Was debating that too. It is the _actual_ dtime as far as the inside of the block is concerned.

There're other places in there where dtime is shadowed in a block. Happy to change this and those too.

SmallJoker commented on src/serverenvironment.cpp:
```diff
@@ -1306,6 +1303,20 @@ void ServerEnvironment::step(float dtime)
+		if (m_avg_ao_time > max_time_ms) {
+			infostream<<"active objects took "
+				<<m_avg_ao_time<<"ms (longer than "
+				<<max_time_ms<<"ms)"<<std::endl;
```
Spaces around operators: `a << b`

lhofhansl commented:
All comments addressed:
* renamed dtime to elapsed_time (including one existing place)
* fixed layout nits (also in existing place)
* turned going over the active limit into a warning (just like it is for ABMs)

paramat commented on src/serverenvironment.cpp:
```diff
@@ -279,17 +279,17 @@ void LBMManager::applyLBMs(ServerEnvironment *env, MapBlock *block, u32 stamp)
 
 void fillRadiusBlock(v3s16 p0, s16 r, std::set<v3s16> &list)
```
Aren't these changes present in the other PR too?

lhofhansl commented:
Can I get an approval @sofar , @sfan5 ? :)

SmallJoker commented on src/serverenvironment.cpp:
```diff
@@ -1235,18 +1232,14 @@ void ServerEnvironment::step(float dtime)
 	}
 
-	if (m_active_block_modifier_interval.step(dtime, m_cache_abm_interval))
+	if (m_active_block_modifier_interval.step(dtime, m_cache_abm_interval * m_active_block_interval_overload_skip))
```
Line too long:
```C++
if (m_active_block_modifier_interval.step(dtime,
		m_cache_abm_interval * m_active_block_interval_overload_skip))
```

SmallJoker replied:
Line too long
```
if (m_active_block_modifier_interval.step(dtime,
		m_cache_abm_interval * m_active_block_interval_overload_skip))
```

SmallJoker commented on src/serverenvironment.cpp:
```diff
@@ -1260,33 +1253,37 @@ void ServerEnvironment::step(float dtime)
-			u32 max_time_ms = 200;
+			const u32 time_ms = timer.stop(true);
+			// allow up to 10% of the budget interval
+			const u32 max_time_ms = m_cache_abm_interval * 1000 / 10;
```
`* 1000.f / 10.f` for type consistency

SmallJoker commented on src/serverenvironment.cpp:
```diff
@@ -1260,33 +1253,37 @@ void ServerEnvironment::step(float dtime)
 		Step active objects
 	*/
-	{
+	if (m_active_object_interval.step(dtime, m_cache_ao_interval * m_active_object_interval_overload_skip)) {
```
Line too long.

SmallJoker commented on src/serverenvironment.cpp:
```diff
@@ -1299,13 +1296,27 @@ void ServerEnvironment::step(float dtime)
+		// calculate a simple moving average
+		m_avg_ao_time = m_avg_ao_time * 0.9f + timer.stop(true) * 0.1f;
+		// allow up to 10% of the budget interval
+		const float max_time_ms = m_cache_ao_interval * 1000 / 10;
```
`* 1000.f / 10.f`

lhofhansl commented:
Thanks @SmallJoker will fix tonight. (can't believe I missed the float constant part again, my brain just seems equate integers and floats.)

lhofhansl commented:
Done ... @SmallJoker

lhofhansl commented:
It seems also slightly better to default the active object step to 0.2s. (instead of 0.1s)

lhofhansl commented:
This should be good now @SmallJoker .
Added correct accumulative time tracking. Gave AOs a 20% budget, since they are a bit more real-time critical than ABMs.

SmallJoker commented on src/serverenvironment.cpp:
```diff
@@ -1194,14 +1191,16 @@ void ServerEnvironment::step(float dtime)
 	*/
-	if (m_active_blocks_nodemetadata_interval.step(dtime, m_cache_nodetimer_interval)) {
+	if (m_active_blocks_nodemetadata_interval.step(dtime, m_cache_nodetimer_interval,
+		&elapsed_time)) {
```
http://dev.minetest.net/Code_style_guidelines#Spaces
> When breaking conditionals indent following lines of the conditional with two tabs and the statement body with one tab. [...]

i.e. Needs an indent (tab) more for conditions that are continued on multiple lines.

lhofhansl commented:
Rebased and fixed indentation.

lhofhansl commented:
9c669016
